doc/reference/vm-live-migration-internals.md

```diff
@@ -0,0 +1,112 @@
+
+For shared storage pools such as Ceph, disk data transfer is unnecessary. Instead, the process focuses on transferring the VM state through a dedicated migration socket.
+
+1. Validate cluster state and storage pool readiness.
+2. Transfer the live state data over the migration socket.
+3. Start the target VM using the transferred state.
+
+## Implementation Details
+
+### Migration API
+
+Sending a `POST` request to `/1.0/instances/{name}` renames, moves an instance between pools or migrates an instance to another server. The returned operation metadata for migration, in the push case, will be a background operation with progress data, for the pull case, it will be a WebSocket operation with a number of secrets to be passed to the target server.
+
+### Live Migration Call Stack
+
+Below is a general overview of the key functions part of the live migration call stack:
+
+#### [`lxd/lxd/instance_post.go`](https://github.com/canonical/lxd/lxd/instance_post.go)
+
+1. [`instancePost`](https://github.com/canonical/lxd/lxd/instance_post.go#L74)
+
+This function handles post requests to the `/1.0/instances` endpoint.
+
+#### [`lxd/lxd/migrate_instance.go`](https://github.com/canonical/lxd/lxd/migrate_instance.go)
+
+1. [`Do`](https://github.com/canonical/lxd/lxd/migrate_instance.go#L87)
+
+This function performs the migration operation on the source VM for the given state and operation. It sets up the necessary WebSocket connections for control, state, and filesystem, and then initiates the migration process.
+
+#### [`lxd/lxd/instance/drivers/driver_qemu.go`](https://github.com/canonical/lxd/lxd/instance/drivers/driver_qemu.go)
+
+1. [`MigrateSend`](https://github.com/canonical/lxd/lxd/instance/drivers/driver_qemu.go#L6436)
+
+This function controls the sending of a migration, checking for stateful support, waiting for connections, performing checks, and sending a migration offer. When performing a intra-cluster same-name migration, steps are taken to prevent corruption of volatile device configuration keys during start and stop of the instance on both source and target.
+
+2. [`migrateSendLive`](https://github.com/canonical/lxd/lxd/instance/drivers/driver_qemu.go#L6666)
+
+This function performs the live migration send process.
+
+##### High-Level Steps
+
+1. Setup connection.
+2. Determine migration type (shared or non-shared storage).
+3. Set migration capabilities.
+4. Non-shared storage preparation.
+- Create and configure snapshot file for root disk writes during migration;
+- Add the snapshot as a block device to the source VM;
+- Redirect disk writes to the snapshot.
+5. Storage transfer.
+- For shared storage, we just perform checks at this point;
+- For non-shared storage, set up an NBD listener and connect it to the target to transfer disk.
+6. Snapshot sync for non-shared storage using `BlockDevMirror`.
+7. Transfer VM state to target.
+
+## Appendix
+
+### Glossary
+
+QEMU: Emulation and virtualization software.
```

Comment:
Note: I would normally advise using the `{abbr}` syntax for inline tooltip-style definitions instead of a glossary, but it seems to be broken currently. Alternatively, we could use the footnote syntax here, if you don't mind the numbering it adds (see https://documentation.ubuntu.com/lxd/en/latest/guest-os-compatibility/ for an example of how it works/looks), or leave it as-is until the `{abbr}` bug is fixed.

Reply:
Tested again - `{abbr}` is working. If you want to use that instead, you can add it like this inline:
```
Create a temporary {abbr}`QCOW2 (QEMU Copy On Write)` file ...
```
This renders as an `<abbr>` tag around `QCOW2` with a `title` attribute of `QEMU Copy On Write`. When a user hovers over `QCOW2`, the text `QEMU Copy On Write` should appear on top. For accessibility reasons, only the expanded form of the abbreviation (`QEMU Copy On Write`) should go in the parentheses, not additional text such as "file format". If you want to keep the additional text, I would advise using the footnote syntax.

Reply:
Thanks for sharing this functionality - I think the `<abbr>` tag will work well in this case!